web/src/stream_list_sort.js

```diff
@@ -68,7 +68,7 @@ export function has_recent_activity(sub) {
         // to set_filter_out_inactives.
         return true;
     }
-    return stream_topic_history.stream_has_topics(sub.stream_id) || sub.newly_subscribed;
+    return all_messages_data.has_message_from_stream(sub.stream_id) || sub.newly_subscribed;
```

Comment:
I think this check is too slow -- since we're walking every message in `all_messages_data` (can be 10ks) for each stream when rerendering the left sidebar.

An alternative that I think would work is to check `stream_topic_history.get_max_message_id(sub_stream_id) >= all_messages_data.first().id` ... which would mean using the oldest message ID in `all_messages_data` as a consistent definition for what qualifies as recent (shared with "Recent conversations"), but avoiding walking that whole data structure.

Reply:
I guess we need to make sure we don't crash if `all_messages_data` contains no messages; an unlikely but theoretically possibly failure mode.

Reply:
A quick update is that this logic doesn't work. After few seconds all the inactive streams move to active section. I'll check the reason.

```
    has_message_from_stream(stream_id) {
        if (this.visibly_empty()) {
            return false;
        }
        return stream_topic_history.get_max_message_id(stream_id) >= this.first().id;
    }
```

Reply:
Hmm, I'd expect there to be some sort of loading order issue, where streams display as inactive early. The left sidebar is rendered using just `page_params`, and we don't have any data in `all_message_data` until after additional `message_fetch` round trips. 

I don't have a good solution for how to fix that ... though I would expect that bug would have been present pre-patch anyway?

I guess one theory is that we only start considering streams as inactive at all once `all_messages_data.fetch_status.has_found_newest` is true, i.e. at the point where we have coherent data?

(In any case, I don't think that being a function on `message_list_data.js`, vs. just a thing in `stream_sort.js` that calls into that, makes a ton of sense).